docs/2-features/17-oauth.md

```diff
@@ -0,0 +1,289 @@
+
+        return new Redirect('/');
+    }
+}
+```
+
+Of course, this example assumes that the database and an [authenticatable model](../2-features/04-authentication.md#authentication) are configured.
+
+### Working with the OAuth user
+
+When an OAuth flow is completed and you call `fetchUser`, you will receive an {b`Tempest\Auth\OAuth\OAuthUser`} object containing the user's information from the OAuth provider:
+
+```php
+$user = $this->oauth->fetchUser($code);
+
+$user->id;         // The unique identifier for the user from the OAuth provider
+$user->email;      // The user's email address
+$user->name;       // The user's name
+$user->nickname;   // The user's nickname/username
+$user->avatar;     // The user's avatar URL
+$user->provider;   // The OAuth provider name
+$user->raw;        // Raw user data from the OAuth provider
+```
+
+As seen in the example above, you can use this information to create or update a user in your database, or to authenticate them directly.
+
+## Configuring a provider
+
+Most providers require only a `clientId`, `clientSecret` and `redirectTo`, but some might need other parameters. A typical configuration looks like the following:
+
+```php app/Auth/github.config.php
+return new GitHubOAuthConfig(
+    clientId: env('GITHUB_CLIENT_ID'),
+    clientSecret: env('GITHUB_CLIENT_SECRET'),
+    redirectTo: [GitHubOAuthController::class, 'callback'],
+    scopes: ['user:email'],
+);
+```
+
+Note that the `redirectTo` accepts a tuple of a controller class and a method name, which will be resolved to the full URL of the route handled by that method. You may also provide an URI path if you prefer.
+
+### Supporting multiple providers
+
+If you need to work with multiple OAuth providers, you may create multiple OAuth configurations using tags. These tags may then be used to resolve the {b`Tempest\Auth\OAuth\OAuthClient`} interface, which will use the corresponding configuration.
+
+It's a good practice to use an enum for the tag:
+
+```php app/Auth/Provider.php
+enum Provider
+{
+    case GITHUB;
+    case GOOGLE;
+    case DISCORD;
+}
+```
+
+```php app/Auth/github.config.php
+return new GitHubOAuthConfig(
+    tag: Provider::GITHUB,
+    clientId: env('GITHUB_CLIENT_ID'),
+    clientSecret: env('GITHUB_CLIENT_SECRET'),
+    redirectTo: [OAuthController::class, 'handleGitHubCallback'],
+    scopes: ['user:email'],
+);
+```
+
+```php app/Auth/google.config.php
+return new GoogleOAuthConfig(
+    tag: Provider::GOOGLE,
+    clientId: env('GOOGLE_CLIENT_ID'),
+    clientSecret: env('GOOGLE_CLIENT_SECRET'),
+    redirectTo: [GoogleOAuthController::class, 'handleGoogleCallback'],
+);
+```
+
+Once you have configured your OAuth providers and your tags, you may inject the {b`Tempest\Auth\OAuth\OAuthClient`} interface using the corresponding tag:
+
+```php app/AuthController.php
+use Tempest\Container\Tag;
+
+final readonly class AuthController
+{
+    public function __construct(
+        #[Tag(OAuthProvider::GITHUB)]
```

Comment:
I JUST WANNA POINT OUT HOW MUCH I LOVE TEMPEST AND HOW COOL IT IS THAT WE'RE ABLE TO DO THIS!!! 🌊